### ISSUE TITLE
`retina_masks=True` will triger incorrect calculation when `pad` in `scale_masks()` is not divisible by 2

### ISSUE THREAD

**Issue by jamjamjon** [ultralytics/ultralytics#6110]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

# Description
When the yolov8-segmentation model sets the `retina_masks=True` for inference,  there is a problem with the masks(segments) position in the output results under certain circumstances(discuss later). The masks(segments) plotted in the image is shifting in a certain direction, like this: 
![upper-clothes-model-wrong](https://github.com/ultralytics/ultralytics/assets/51357717/4048261b-5404-4d58-b338-7d9a79980878)

mask of upper-clothes is shifting to the right.

# Bug reason
The root of the problem is the calculations in `scale_masks()` function. https://github.com/ultralytics/ultralytics/blob/ff921ca93ed2c7ca6d16a199f71b36707fd426f8/ultralytics/utils/ops.py#L691-L721

When the `pad` variable cannot be evenly divided by 2, mistake happened. 
https://github.com/ultralytics/ultralytics/blob/ff921ca93ed2c7ca6d16a199f71b36707fd426f8/ultralytics/utils/ops.py#L703-L706

Using `int()` directlly will round down the `pad[0]` and `pad[1]` and cause one of `top`, `left`, `bottom` and `right` to be calculated less than the correct value. 
https://github.com/ultralytics/ultralytics/blob/ff921ca93ed2c7ca6d16a199f71b36707fd426f8/ultralytics/utils/ops.py#L707-L708

Besides,
 `scale_masks()` for masks,  [https://github.com/ultralytics/ultralytics/blob/ff921ca93ed2c7ca6d16a199f71b36707fd426f8/ultralytics/utils/ops.py#L691](url)
 `scale_boxes()` for bboxes, [https://github.com/ultralytics/ultralytics/blob/ff921ca93ed2c7ca6d16a199f71b36707fd426f8/ultralytics/utils/ops.py#L83](url)
They are not calculated consistently, and the bboxes re-scaled will be used in the subsequent `crop_mask()` later, that's why the outputs' pixels shifing happened.

# Solution
**Modify from**
https://github.com/ultralytics/ultralytics/blob/ff921ca93ed2c7ca6d16a199f71b36707fd426f8/ultralytics/utils/ops.py#L707-L708

**to**
```
    top, left = (int(round(pad[1] - 0.1)), int(round(pad[0] - 0.1))) if padding else (0, 0)  # y, x
    bottom, right = (int(round(mh - pad[1] + 0.1)), int(round(mw - pad[0] + 0.1)))
```

### then mistake rectified

![upper-clothes-model-fixed](https://github.com/ultralytics/ultralytics/assets/51357717/ad2cc7ae-9a17-4333-a8bd-ce3401d79f48)

### Environment

Ultralytics YOLOv8.0.205 🚀 Python-3.10.13 torch-2.1.0+cu121 CPU (11th Gen Intel Core(TM) i7-11700 2.50GHz)
python3.10.13
ubuntu18.04

## Source image
![image-test](https://github.com/ultralytics/ultralytics/assets/51357717/9e9b11a8-54d1-4e51-b843-ce38231e6138)

## Models (.pt has the same results)
- yolov8s-seg.onnx
- yolov8l-seg.onnx
- upper-clothes.onnx(imgsz=(384，384)) (file size too big, can not upload)

```
yolo predict task=segment imgsz=640 model=yolov8s-seg.onnx source=image-test.jpg classes=0 retina_masks=true
yolo predict task=segment imgsz=640 model=yolov8l-seg.onnx source=image-test.jpg classes=0 retina_masks=true
# yolo predict task=segment imgsz=384 model=upper-clothes.onnx source=image-test.jpg classes=0 retina_masks=true
```

## Results of using current ultralytics main branch
![upper-clothes-model-wrong](https://github.com/ultralytics/ultralytics/assets/51357717/1031c74e-8c14-4f48-86be-958b8b278c44)
![yolov8l-seg-wrong](https://github.com/ultralytics/ultralytics/assets/51357717/3d9bca1a-ce07-4bbf-b29e-f257209039b2)
![yolov8s-seg-wrong](https://github.com/ultralytics/ultralytics/assets/51357717/d9c1ab5e-4bb3-4f64-b620-0ff10296616b)

## Results after modifying ( above solution)
```
def scale_masks(masks, shape, padding=True):
    ...
    # ---> original 
    # top, left = (int(pad[1]), int(pad[0])) if padding else (0, 0)  # y, x
    # bottom, right = (int(mh - pad[1]), int(mw - pad[0]))

    # ---> modified 
    top, left = (int(round(pad[1] - 0.1)), int(round(pad[0] - 0.1))) if padding else (0, 0)  # y, x
    bottom, right = (int(round(mh - pad[1] + 0.1)), int(round(mw - pad[0] + 0.1)))
    ...

```
![upper-clothes-model-fixed](https://github.com/ultralytics/ultralytics/assets/51357717/06efe7f3-0e41-4319-9165-fe8559a3008c)
![yolov8l-seg-fixed](https://github.com/ultralytics/ultralytics/assets/51357717/2c9a7f09-1970-407a-bb41-064e583e7e25)
![yolov8s-seg-fixed](https://github.com/ultralytics/ultralytics/assets/51357717/c2ea768c-e485-44be-9a12-a035954e1b43)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@jamjamjon thank you for your detailed report and for your proactive solution. We appreciate your collaboration in maintaining and improving the YOLOv8's performance and reliability.

The issue you raise regarding the `scale_masks()` function is a meaningful one. The function's current configuration does round down the `pad[0]` and `pad[1]` which indeed could cause a shift in the calculated values of `top`, `left`, `bottom`, and `right`. This can result in the masks being misaligned, which affects the detection accuracy of our model. 

Your proposal of using the rounding instead of the direct integer casting in the `scale_masks()` function's statement could indeed rectify this calculation error. Also, the reasoning to bring the consistency between the `scale_masks()` and `scale_boxes()` computations is logical, and might provide a substantial improvement in the detection accuracy.

We will review your suggestion and explore its implications. If your solution proves to be reliable, we will implement it in the next YOLOv8 update.

Again, we appreciate your contribution to the ongoing development of YOLOv8. Your willingness to submit a Pull Request is excellent news. Please use our contributing guidelines for your PR. We look forward to collaborating with you on this improvement.

**Comment by jamjamjon** [comment]
I made a PR #6121

**Comment by glenn-jocher** [comment]
@jamjamjon great news 😃! Your original issue may now be resolved ✅ in PR #6121. This PR should publish with `ultralytics 8.0.206` shortly.

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
